File: serp_llm/src/core/mapper/message_mapper.py
"""
Author: QuanTuanHuy
Description: Part of Serp Project - Message Mapper
"""

from typing import Optional

from src.core.domain.entities import MessageEntity
from src.core.domain.enums import MessageRole, ContentType
from src.infrastructure.db.models import MessageModel


class MessageMapper:
    """Mapper between MessageEntity and MessageModel"""
    
    @staticmethod
    def to_entity(model: Optional[MessageModel]) -> Optional[MessageEntity]:
        """Convert database model to domain entity"""
        if model is None:
            return None
        
        return MessageEntity(
            id=model.id,
            conversation_id=model.conversation_id,
            role=MessageRole(model.role),
            content=model.content,
            content_type=ContentType(model.content_type),
            attachments=model.attachments or [],
            function_call=model.function_call,
            tokens_used=model.tokens_used,
            model_used=model.model_used,
            processing_time_ms=model.processing_time_ms,
            sources=model.sources or [],
            meta_data=model.meta_data or {},
            created_at=model.created_at,
        )
    
    @staticmethod
    def to_model(entity: MessageEntity) -> MessageModel:
        """Convert domain entity to database model"""
        return MessageModel(
            id=entity.id,
            conversation_id=entity.conversation_id,
            role=entity.role.value,
            content=entity.content,
            content_type=entity.content_type.value,
            attachments=entity.attachments,
            function_call=entity.function_call,
            tokens_used=entity.tokens_used,
            model_used=entity.model_used,
            processing_time_ms=entity.processing_time_ms,
            sources=entity.sources,
            meta_data=entity.meta_data,
            created_at=entity.created_at,
        )

File: discuss_service/src/main/java/serp/project/discuss_service/infrastructure/store/mapper/MessageMapper.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project - Message entity/model mapper
 */

package serp.project.discuss_service.infrastructure.store.mapper;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Component;
import serp.project.discuss_service.core.domain.entity.MessageEntity;
import serp.project.discuss_service.infrastructure.store.model.MessageModel;

import java.util.*;
import java.util.stream.Collectors;

@Component
public class MessageMapper extends BaseMapper {

    private final ObjectMapper objectMapper = new ObjectMapper();

    public MessageEntity toEntity(MessageModel model) {
        if (model == null) {
            return null;
        }

        return MessageEntity.builder()
                .id(model.getId())
                .channelId(model.getChannelId())
                .senderId(model.getSenderId())
                .tenantId(model.getTenantId())
                .content(model.getContent())
                .messageType(model.getMessageType())
                .mentions(arrayToList(model.getMentions()))
                .parentId(model.getParentId())
                .threadCount(model.getThreadCount())
                .isEdited(model.getIsEdited())
                .editedAt(localDateTimeToLong(model.getEditedAt()))
                .isDeleted(model.getIsDeleted())
                .deletedAt(localDateTimeToLong(model.getDeletedAt()))
                .deletedBy(model.getDeletedBy())
                .reactions(parseReactions(model.getReactions()))
                .metadata(model.getMetadata())
                .createdAt(localDateTimeToLong(model.getCreatedAt()))
                .updatedAt(localDateTimeToLong(model.getUpdatedAt()))
                .build();
    }

    public MessageModel toModel(MessageEntity entity) {
        if (entity == null) {
            return null;
        }

        return MessageModel.builder()
                .id(entity.getId())
                .channelId(entity.getChannelId())
                .senderId(entity.getSenderId())
                .tenantId(entity.getTenantId())
                .content(entity.getContent())
                .messageType(entity.getMessageType())
                .mentions(listToArray(entity.getMentions()))
                .parentId(entity.getParentId())
                .threadCount(entity.getThreadCount())
                .isEdited(entity.getIsEdited())
                .editedAt(longToLocalDateTime(entity.getEditedAt()))
                .isDeleted(entity.getIsDeleted())
                .deletedAt(longToLocalDateTime(entity.getDeletedAt()))
                .deletedBy(entity.getDeletedBy())
                .reactions(serializeReactions(entity.getReactions()))
                .metadata(entity.getMetadata())
                .createdAt(longToLocalDateTime(entity.getCreatedAt()))
                .updatedAt(longToLocalDateTime(entity.getUpdatedAt()))
                .build();
    }

    public List<MessageEntity> toEntityList(List<MessageModel> models) {
        if (models == null) {
            return null;
        }
        return models.stream()
                .map(this::toEntity)
                .collect(Collectors.toList());
    }

    public List<MessageModel> toModelList(List<MessageEntity> entities) {
        if (entities == null) {
            return null;
        }
        return entities.stream()
                .map(this::toModel)
                .collect(Collectors.toList());
    }

    // Array <-> List conversion
    private List<Long> arrayToList(Long[] array) {
        if (array == null) {
            return new ArrayList<>();
        }
        return Arrays.asList(array);
    }

    private Long[] listToArray(List<Long> list) {
        if (list == null) {
            return new Long[0];
        }
        return list.toArray(new Long[0]);
    }

    // Reaction conversion: List<ReactionVO> <-> List<Map<String, Object>>
    private List<MessageEntity.ReactionVO> parseReactions(List<Map<String, Object>> reactionsJson) {
        if (reactionsJson == null || reactionsJson.isEmpty()) {
            return new ArrayList<>();
        }

        List<MessageEntity.ReactionVO> reactions = new ArrayList<>();
        for (Map<String, Object> r : reactionsJson) {
            String emoji = (String) r.get("emoji");
            @SuppressWarnings("unchecked")
            List<Number> userIdNumbers = (List<Number>) r.get("userIds");
            List<Long> userIds = new ArrayList<>();
            if (userIdNumbers != null) {
                for (Number n : userIdNumbers) {
                    userIds.add(n.longValue());
                }
            }
            reactions.add(new MessageEntity.ReactionVO(emoji, userIds));
        }
        return reactions;
    }

    private List<Map<String, Object>> serializeReactions(List<MessageEntity.ReactionVO> reactions) {
        if (reactions == null || reactions.isEmpty()) {
            return new ArrayList<>();
        }

        List<Map<String, Object>> result = new ArrayList<>();
        for (MessageEntity.ReactionVO r : reactions) {
            Map<String, Object> map = new HashMap<>();
            map.put("emoji", r.getEmoji());
            map.put("userIds", r.getUserIds());
            result.add(map);
        }
        return result;
    }
}
